bindings/rust/bench/src/bin/graph_perf.rs

```diff
@@ -0,0 +1,306 @@
+            0.0..(1.2 * y_max),
+        )
+        .unwrap();
+
+    let axis_label_style = ("sans-serif", 18).into_font();
+
+    ctx.configure_mesh()
+        .light_line_style(RGBAColor(235, 235, 235, 1.0)) // gridline color
+        .bold_line_style(RGBAColor(225, 225, 225, 1.0))
+        .x_desc("Version") // axis labels
+        .x_labels(20) // max number of labels
+        .x_label_style(axis_label_style.clone())
+        .x_label_formatter(x_label_formatter)
+        .y_desc(y_label)
+        .y_labels(5)
+        .y_label_formatter(y_label_formatter)
+        .y_label_style(axis_label_style)
+        .draw()
+        .unwrap();
+
+    // go through each DataSeries and plot them
+    for (i, data_series) in data.iter().enumerate() {
+        // remove data that returned error while benching
+        // heuristic: times < 1% of y_max are invalid/had error
+        let filtered_data = data_series
+            .data
+            .iter()
+            .filter(|data_point| data_point.y > 0.01 * y_max)
+            .collect::<Vec<_>>();
+
+        let color = Palette99::pick(i);
+
+        // draw error bars
+        ctx.draw_series(filtered_data.iter().map(|data_point| {
+            ErrorBar::new_vertical(
+                data_point.x,
+                data_point.y - data_point.y_bar,
+                data_point.y,
+                data_point.y + data_point.y_bar,
+                &color,
+                3,
+            )
+        }))
+        .unwrap();
+
+        // draw lines with legend entry
+        ctx.draw_series(LineSeries::new(
+            filtered_data
+                .iter()
+                .map(|data_point| (data_point.x, data_point.y)),
+            color.stroke_width(2),
+        ))
+        .unwrap()
+        .label(&data_series.name)
+        .legend(move |(x, y)| Rectangle::new([(x, y - 5), (x + 10, y + 5)], color.filled()));
+    }
+
+    // enable legend
+    ctx.configure_series_labels()
+        .position(SeriesLabelPosition::LowerRight)
+        .margin(10)
+        .border_style(BLACK)
+        .background_style(WHITE)
+        .draw()
+        .unwrap();
+}
+
+fn main() {
+    let handshake_data = get_all_data("handshake");
+    let throughput_data = get_all_data("throughput");
+
+    // combine all versions present in handshake and throughput data
+    // also fill in missing version v1.3.15 and v1.3.30-v1.3.37
+    let versions = get_unique_versions(&handshake_data)
+        .into_iter()
+        .chain(get_unique_versions(&throughput_data).into_iter())
+        .chain((15..16).chain(30..38).map(|p| Version::new(1, 3, p)))
+        .collect::<BTreeSet<Version>>()
+        .into_iter()
+        .collect::<Vec<Version>>();
+
+    // map versions to x coordinates
+    let version_to_x = versions
+        .iter()
+        .enumerate()
+        .map(|(i, version)| (version, i as i32))
+        .collect::<HashMap<&Version, i32>>();
```

Comment:
nit: wonder if its possible to simplify this. Try to avoid doing `.collect` more than once, since that allocates. The cool thing about operations on iter (map, chain, filter) is that they are lazy.. meaning that it doest create a collection at each step. A collection is only created when you do `collect` so its ideal to have only one.

Reply:
I chained the `collect::<BTreeSet<Version>>()` and `collect::Vec<Version>()` calls because I need to sort and remove duplicates from the iterator, but I think extending a set from one `get_unique_versions()` call would be better; just changed it.